File: 2018_MotionProfiling/src/Spline.java

public class Spline {

    public static class Point {
        private double x;
        private double y;
        private double heading;

        public Point(double x, double y, double heading) {
            this.x = x;
            this.y = y;
            this.heading = heading;
        }

        public Point(double x, double y) {
            this(x, y, 999999);
        }

        public double getX() {
            return this.x;
        }

        public double getY() {
            return this.y;
        }

        public double getHeading() {
            return this.heading;
        }
    }

    private Point[] points;

    private Polynomial[] xPolynomials;
    private Polynomial[] yPolynomials;

    private Polynomial[] xFirstDerivatives;
    private Polynomial[] yFirstDerivatives;

    private Polynomial[] xSecondDerivatives;
    private Polynomial[] ySecondDerivatives;

    public Spline(Point[] points, Polynomial[] xPolynomials, Polynomial[] yPolynomials) {

        assert xPolynomials.length == yPolynomials.length : "xPolynomials and yPolynomials must be the same length.";
        this.points = points;
        this.xPolynomials = xPolynomials;
        this.yPolynomials = yPolynomials;

        this.xFirstDerivatives = new Polynomial[xPolynomials.length];
        this.yFirstDerivatives = new Polynomial[yPolynomials.length];

        this.xSecondDerivatives = new Polynomial[xPolynomials.length];
        this.ySecondDerivatives = new Polynomial[yPolynomials.length];

        for (int i = 0; i < xPolynomials.length; i++) {
            this.xFirstDerivatives[i] = Polynomial.differentiate(this.xPolynomials[i]);
            this.yFirstDerivatives[i] = Polynomial.differentiate(this.yPolynomials[i]);
            this.xSecondDerivatives[i] = Polynomial.differentiate(this.xFirstDerivatives[i]);
            this.ySecondDerivatives[i] = Polynomial.differentiate(this.yFirstDerivatives[i]);
        }
    }

    public Point evaluate(double t) {
        if (t == this.xPolynomials.length) {
            return new Point(this.xPolynomials[this.xPolynomials.length - 1].evaluate(1.0),
                    this.yPolynomials[this.yPolynomials.length - 1].evaluate(1.0));
        } else {
            int index = (int) t;
            return new Point(xPolynomials[index].evaluate(t - index), yPolynomials[index].evaluate(t - index));
        }
    }

    public Point evaluateFirstDerivatives(double t) {
        if (t == this.xFirstDerivatives.length) {
            return new Point(this.xFirstDerivatives[this.xFirstDerivatives.length - 1].evaluate(1.0),
                    this.yFirstDerivatives[this.yFirstDerivatives.length - 1].evaluate(1.0));
        } else {
            int index = (int) t;
            return new Point(xFirstDerivatives[index].evaluate(t - index),
                    yFirstDerivatives[index].evaluate(t - index));
        }
    }

    public Point evaluateSecondDerivatives(double t) {
        if (t == this.xSecondDerivatives.length) {
            return new Point(this.xSecondDerivatives[this.xSecondDerivatives.length - 1].evaluate(1.0),
                    this.ySecondDerivatives[this.ySecondDerivatives.length - 1].evaluate(1.0));
        } else {
            int index = (int) t;
            return new Point(xSecondDerivatives[index].evaluate(t - index),
                    ySecondDerivatives[index].evaluate(t - index));
        }
    }

    public double evaluateHeading(double t) {
//        double y = evaluate(0).y;
//        double x = evaluate(0).x;
//        double initialHeading = Math.toDegrees(Math.atan2(y, x));
        double xDistance = evaluateFirstDerivatives(t).x;
        double yDistance = evaluateFirstDerivatives(t).y;

        double heading = Math.toDegrees(Math.atan2(yDistance, xDistance));

        if (heading < 0) {
            heading += 360.0;
        }
//        double correctedHeading = (heading - initialHeading) > 0.0 ? heading - initialHeading
//                : heading - initialHeading + 360.0;

        return heading;
    }


    public double evaluateCurvature(double t) {
        Point firstDerivative = evaluateFirstDerivatives(t);
        Point secondDerivative = evaluateSecondDerivatives(t);

        return (firstDerivative.getX() * secondDerivative.getY() - firstDerivative.getY() * secondDerivative.getX()) // (x'y'-
                // y'x'')
                / Math.pow(Math.pow(firstDerivative.getX(), 2) + Math.pow(firstDerivative.getY(), 2), 1.5); // / (x'^2)
        // y'^2)^1.5
    }

    public Point[] getPoints() {
        return this.points;
    }

    public double getDomainMax() {
        return this.xPolynomials.length;
    }

    public String pointsToString() {
        String pointsString = "";
        for (Point point : this.points) {
            pointsString += "(" + point.getX() + "," + point.getY() + "),";
        }

        return pointsString.substring(0, pointsString.length() - 1);
    }

    public String curvatureToDesmos() {
        String equation = "\\left\\{";
        for (int i = 0; i < this.xPolynomials.length; i++) {
            String parameterString = "(x-" + i + ")";
            String curvatureEquation = "((" + this.xFirstDerivatives[i].toString(parameterString) + ")*("
                    + this.ySecondDerivatives[i].toString(parameterString) + ")-("
                    + this.yFirstDerivatives[i].toString(parameterString) + ")*("
                    + this.xSecondDerivatives[i].toString(parameterString) + "))/((("
                    + this.xFirstDerivatives[i].toString(parameterString) + "))^2+(("
                    + this.yFirstDerivatives[i].toString(parameterString) + "))^2)^{\\frac{3}{2}}";

            equation += i + "<=x<" + (i + 1) + ":" + curvatureEquation + ",";
        }

        return equation.substring(0, equation.length() - 1) + "\\right\\}";
    }

    public String toString() {
        return "(" + polynomialListToString(this.xPolynomials) + ", " + polynomialListToString(this.yPolynomials) + ")";
    }

    public String firstDerivativeToString() {
        return "(" + polynomialListToString(this.xFirstDerivatives) + ", "
                + polynomialListToString(this.yFirstDerivatives) + ")";
    }

    public String secondDerivativeToString() {
        return "(" + polynomialListToString(this.xSecondDerivatives) + ", "
                + polynomialListToString(this.ySecondDerivatives) + ")";
    }

    private String polynomialListToString(Polynomial[] polynomials) {
        String equation = "\\left\\{";
        for (int i = 0; i < polynomials.length; i++) {
            equation += i + "<=t<" + (i + 1) + ":" + polynomials[i].toString("(t-" + i + ")") + ",";
        }

        return equation.substring(0, equation.length() - 1) + "\\right\\}";
    }

    public static Spline quinticHermiteSplineFor(double velocityWeight, double accelerationWeight, Point... points) {

        Polynomial[] xPolynomials = QuinticHermite.splineFor(velocityWeight, accelerationWeight, points, true);
        Polynomial[] yPolynomials = QuinticHermite.splineFor(velocityWeight, accelerationWeight, points, false);

        return new Spline(points, xPolynomials, yPolynomials);

    }
}

File: TrajectoryGeneration/trajectory/Spline.go
package trajectory

import (
	"fmt"
	"math"
)

type Point struct {
	X float64 `json: "X"`
	Y float64 `json: "Y"`
}

type Point2 struct {
	X, Y    float64
	Heading float64
	Tension float64
}

func (p Point2) Point() Point {
	return Point{p.X, p.Y}
}

type Spline struct {
	points                                 []Point
	xPolynomials, yPolynomials             []Polynomial
	xFirstDerivatives, yFirstDerivatives   []Polynomial
	xSecondDerivatives, ySecondDerivatives []Polynomial
}

func NewSpline(points []Point, xPolynomials, yPolynomials []Polynomial) *Spline {
	xFirstDerivatives := make([]Polynomial, len(xPolynomials))
	yFirstDerivatives := make([]Polynomial, len(yPolynomials))
	xSecondDerivatives := make([]Polynomial, len(xPolynomials))
	ySecondDerivatives := make([]Polynomial, len(yPolynomials))

	for i := 0; i < len(xPolynomials); i++ {
		xFirstDerivatives[i] = Differentiate(xPolynomials[i])
		yFirstDerivatives[i] = Differentiate(yPolynomials[i])
		xSecondDerivatives[i] = Differentiate(xFirstDerivatives[i])
		ySecondDerivatives[i] = Differentiate(yFirstDerivatives[i])
	}

	return &Spline{points, xPolynomials, yPolynomials, xFirstDerivatives, yFirstDerivatives, xSecondDerivatives, ySecondDerivatives}
}

func (spline *Spline) Evaluate(t float64) Point {
	if int(t) == len(spline.xPolynomials) {
		return Point{spline.xPolynomials[len(spline.xPolynomials)-1].Evaluate(1.0), spline.yPolynomials[len(spline.yPolynomials)-1].Evaluate(1.0)}
	} else {
		index := math.Floor(t)
		return Point{spline.xPolynomials[int(index)].Evaluate(t - index), spline.yPolynomials[int(index)].Evaluate(t - index)}
	}
}

func (spline *Spline) EvaluateFirstDerivative(t float64) Point {
	if int(t) == len(spline.xFirstDerivatives) {
		return Point{spline.xFirstDerivatives[len(spline.xFirstDerivatives)-1].Evaluate(1.0), spline.yFirstDerivatives[len(spline.yFirstDerivatives)-1].Evaluate(1.0)}
	} else {
		index := math.Floor(t)
		return Point{spline.xFirstDerivatives[int(index)].Evaluate(t - index), spline.yFirstDerivatives[int(index)].Evaluate(t - index)}
	}
}

func (spline *Spline) EvaluateSecondDerivative(t float64) Point {
	if int(t) == len(spline.xSecondDerivatives) {
		return Point{spline.xSecondDerivatives[len(spline.xSecondDerivatives)-1].Evaluate(1.0), spline.ySecondDerivatives[len(spline.ySecondDerivatives)-1].Evaluate(1.0)}
	} else {
		index := math.Floor(t)
		return Point{spline.xSecondDerivatives[int(index)].Evaluate(t - index), spline.ySecondDerivatives[int(index)].Evaluate(t - index)}
	}
}

func (spline *Spline) EvaluateHeading(t float64) float64 {
	firstDerivative := spline.EvaluateFirstDerivative(t)
	return math.Atan2(firstDerivative.Y, firstDerivative.X) * 180 / math.Pi
}

func (spline *Spline) EvaluateCurvature(t float64) float64 {
	firstDerivative := spline.EvaluateFirstDerivative(t)
	secondDerivative := spline.EvaluateSecondDerivative(t)

	return (firstDerivative.X*secondDerivative.Y - firstDerivative.Y*secondDerivative.X) / math.Pow(math.Pow(firstDerivative.X, 2)+math.Pow(firstDerivative.Y, 2), 1.5)
}

func (spline *Spline) DomainMax() float64 {
	return float64(len(spline.xPolynomials))
}

func (spline Spline) String() string {
	return "(" + PolynomialListToString(spline.xPolynomials) + "," + PolynomialListToString(spline.yPolynomials) + ")"
}

func PolynomialListToString(polynomials []Polynomial) string {
	equation := "\\left\\{"
	for index, polynomial := range polynomials {
		equation += fmt.Sprintf("%d<=t<%d:%s,", index, index+1, polynomial.StringForParameter(fmt.Sprintf("(t-%d)", index)))
	}

	return string([]rune(equation)[:len(equation)-1]) + "\\right\\}"
}

func SplineFor(startHeadingDegrees, endHeadingDegrees float64, points ...Point) *Spline {
	startHeading := startHeadingDegrees * math.Pi / 180
	endHeading := endHeadingDegrees * math.Pi / 180

	startDistance := math.Sqrt(math.Pow(points[1].X-points[0].X, 2) + math.Pow(points[1].Y-points[0].Y, 2))
	endDistance := math.Sqrt(math.Pow(points[len(points)-1].X-points[len(points)-2].X, 2) + math.Pow(points[len(points)-1].Y-points[len(points)-2].Y, 2))

	xValues := make([]float64, len(points))
	yValues := make([]float64, len(points))
	for index, point := range points {
		xValues[index] = point.X
		yValues[index] = point.Y
	}

	xPolynomials := QuinticHermiteFor(0.5*math.Cos(startHeading)*startDistance, 0.5*math.Cos(endHeading)*endDistance, xValues...)
	yPolynomials := QuinticHermiteFor(0.5*math.Sin(startHeading)*startDistance, 0.5*math.Sin(endHeading)*endDistance, yValues...)

	return NewSpline(points, xPolynomials, yPolynomials)
}

func SplineFor2(points ...Point2) *Spline {
	xValues := make([]float64, len(points))
	xVValues := make([]float64, len(points))
	yValues := make([]float64, len(points))
	yVValues := make([]float64, len(points))
	for index, point := range points {
		xValues[index] = point.X
		yValues[index] = point.Y
	}
	for index, point := range points {
		var distanceX, distanceY float64
		if index == 0 {
			distanceX = 2 * math.Abs(xValues[index+1]-xValues[index])
			distanceY = 2 * math.Abs(yValues[index+1]-yValues[index])
		} else if index == len(points)-1 {
			distanceX = 2 * math.Abs(xValues[index]-xValues[index-1])
			distanceY = 2 * math.Abs(yValues[index]-yValues[index-1])
		} else {
			distanceX = math.Abs(xValues[index+1] - xValues[index-1])
			distanceY = math.Abs(yValues[index+1] - yValues[index-1])
		}

		distance := math.Sqrt(math.Pow(distanceX, 2) + math.Pow(distanceY, 2))

		xVValues[index] = (1 - point.Tension) * distance * math.Cos(point.Heading*math.Pi/180)
		yVValues[index] = (1 - point.Tension) * distance * math.Sin(point.Heading*math.Pi/180)
	}

	xPolynomials := QuinticHermiteFor2(xValues, xVValues)
	yPolynomials := QuinticHermiteFor2(yValues, yVValues)

	simplePoints := make([]Point, len(points))
	for index, point := range points {
		simplePoints[index] = point.Point()
	}

	return NewSpline(simplePoints, xPolynomials, yPolynomials)
}

func SplineForSkeleton(skeleton *SplineSkeleton) *Spline {
	length := len(skeleton.X.Values)
	xPolynomials := make([]Polynomial, length-1)
	yPolynomials := make([]Polynomial, length-1)
	for i := 0; i < length-1; i++ {
		xPolynomials[i] = NewQuinticHermite(skeleton.X.Values[i], skeleton.X.Values[i+1], skeleton.X.VValues[i], skeleton.X.VValues[i+1], skeleton.X.AValues[i], skeleton.X.AValues[i+1])
		yPolynomials[i] = NewQuinticHermite(skeleton.Y.Values[i], skeleton.Y.Values[i+1], skeleton.Y.VValues[i], skeleton.Y.VValues[i+1], skeleton.Y.AValues[i], skeleton.Y.AValues[i+1])
	}

	points := make([]Point, length)
	for i := 0; i < length; i++ {
		points[i] = Point{skeleton.X.Values[i], skeleton.Y.Values[i]}
	}

	return NewSpline(points, xPolynomials, yPolynomials)
}
